source/common/router/router.cc

```diff
@@ -380,8 +396,10 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
 
   ENVOY_STREAM_LOG(debug, "router decoding headers:\n{}", *callbacks_, headers);
 
-  upstream_request_ = std::make_unique<UpstreamRequest>(*this, *conn_pool);
-  upstream_request_->encodeHeaders(end_stream);
+  UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(*this, *conn_pool);
+  UpstreamRequest* upstream_request_ptr = upstream_request.get();
+  upstream_request->moveIntoList(std::move(upstream_request), upstream_requests_);
+  upstream_request_ptr->encodeHeaders(end_stream);
```

Comment:
Instead of using a temp variable you could just do `upstream_requests_.front()->encodeHeaders(end_stream);`